Add option to nest context within the h1 for title component

jon-kirwan opened the pull request:
## What
Move the caption inside of the h1.

## Why
Currently, the context is missed when page titles are read by screen readers. This way when users navigate through headings with a screenreader will hear the title with context announced.

No visual changes.

https://trello.com/c/gPFqiwZa/799-update-headings-for-smart-answers-outcome

andysellick commented on app/views/govuk_publishing_components/components/_title.html.erb:
```diff
@@ -22,12 +22,12 @@
     </span>
-  <% end %>
-  <h1 class="<%= heading_classes.join(" ") %>">
+    <% end %>
```
Nitpick: the span inside this `if` statement should be indented.

andysellick reviewed (approved):
Looks good. One small thing - the PR description doesn't explicitly state which component this change refers to, I had to figure it out from the files changed.

maxgds reviewed (commented):
I feel like this ability to put the context within the H1 should be an option rather than the default. The design system says "If the caption should be considered part of the page heading, you can also nest the caption within the h1." and I'm not sure that we want to retroactively update all instances of it that are on the site already - but I'm not sure that we don't, either.  I'm going to open a discussion on the govuk frontenders Slack channel.

jon-kirwan commented:
Thanks all. @andysellick. I've updated the PR to indicate the title component has been updated and have amended indentation in the template too.

@maxgds - i've now added this as an option instead so it can be implemented like so -

```ruby
<%= render "govuk_publishing_components/components/title", {
  context: "Publication",
  title: "My page title",
  context_inside: true
} %>
```

I've also updated the docs and tests.

I'm still quite new to Ruby so would appreciate any feedback on a better way to reference the context (which is currently repeated for both context options) -

```ruby
<span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
  <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
</span>
```

maxgds commented:
Looking good. I can see there are going to be some conflicts with the code I updated yesterday which removes the context href, so you're probably going to need a rebase against master and might have to untangle that a bit, sorry.

In terms of the repetitive context, you could use a content_for block to define it once in the template, see docs here:
https://apidock.com/rails/ActionView/Helpers/CaptureHelper/content_for

I'd also suggest a tweak to your tests - in each one checking for context I'd add a negative assertion for the other position - so we know it just appears once. Docs for that are here: https://apidock.com/rails/ActionController/Assertions/SelectorAssertions/assert_select

andysellick commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@
 
+## Unreleased
+
+* Add option to nest the context within the `<h1>` ([PR #2226](https://github.com/alphagov/govuk_publishing_components/pull/2226))
```
This should match the PR title.

andysellick commented on Gemfile.lock:
```diff
@@ -74,7 +74,7 @@ GEM
       public_suffix (>= 2.0.2, < 5.0)
     ast (2.4.2)
-    brakeman (5.0.4)
+    brakeman (5.1.1)
```
Changes to dependencies in `Gemfile.lock` shouldn't be included in a PR to change a component unless those dependencies are related to the change - which in this case I think they're not. Can these changes be removed?

andysellick commented on app/views/govuk_publishing_components/components/_title.html.erb:
```diff
@@ -21,13 +22,20 @@
   <h1 class="<%= heading_classes.join(" ") %>">
+    <% if context && context_inside %>
+      <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
+        <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
```
It looks like these lines are the same as lines 28-30? In which case you can use the rails `capture` method, something like this:

```
<% context_block = capture do %>
    <span class="govuk-caption-xl gem-c-title__context" <%= "lang=#{context_locale}" if context_locale.present? %>>
      <%= context_href ? link_to(context_text, context_href, class: 'gem-c-title__context-link govuk-link', data: context_data) : context_text %>
    </span>
<% end %>

...

<% if context && !context_inside %>
  <%= context_block %> 
<% end %>

<h1>
  <% if context &&  context_inside %>
    <%= context_block %>
  <% end %>
</h1>

```

andysellick commented on app/views/govuk_publishing_components/components/docs/title.yml:
```diff
@@ -24,6 +24,13 @@ examples:
       title: My page title
+  with_context_inside:
+    description: |
+      If the context should be considered part of the page heading, you can also nest the context within the <code><h1></code>.
```
Nitpick: 'also' is unnecessary, `...you can nest the context within...`

andysellick commented on spec/components/title_spec.rb:
```diff
@@ -21,6 +21,11 @@ def component_name
     assert_select ".govuk-caption-xl", text: "Format"
   end
 
+  it "title context inside appears" do
```
`title context appears inside` maybe?

jon-kirwan commented:
Thanks both. All changes above have now been added / amended.

maxgds reviewed (approved):
Looks good to me!
